cosmos/precompile/erc20/middleware.go

```diff
@@ -0,0 +1,176 @@
+	resp, err := c.em.ERC20AddressForCoinDenom(
+		ctx, &erc20types.ERC20AddressForCoinDenomRequest{
+			Denom: denom,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	var token common.Address
+	if resp.Token == "" {
+		// first occurrence of an IBC originated SDK coin, must be created as a Polaris ERC20 token
+
+		// deploy the new ERC20 token contract (deployer of this contract is the ERC20 module!)
+		if token, err = c.deployPolarisERC20Contract(sdkCtx, evm, c.RegistryKey(), denom, value); err != nil {
+			return err
+		}
+
+		// create the new ERC20 token contract pairing with SDK coin denomination
+		c.em.RegisterCoinERC20Pair(sdkCtx, denom, token)
+	} else {
+		// subsequent occurrence of an IBC-originated SDK coin OR an ERC20 originated token's
+		// Polaris coin counterpart
+
+		// convert ERC20 token bech32 address to common.Address
+		var tokenAcc sdk.AccAddress
+		if tokenAcc, err = sdk.AccAddressFromBech32(resp.Token); err != nil {
+			return err
+		}
+		token = cosmlib.AccAddressToEthAddress(tokenAcc)
+	}
+
+	if erc20types.IsPolarisDenom(denom) {
+		// converting Polaris coins to ERC20 originated tokens
+		// NOTE: it is guaranteed that the ERC20 tokens were transferred to the ERC20 module
+		// precompile contract as escrow before this case is reached.
+
+		// transfer amount ERC20 tokens to the owner
+		if err = c.callERC20Transfer(sdkCtx, evm, c.RegistryKey(), token, owner, amount); err != nil {
+			return err
+		}
+	} else {
+		// converting IBC-originated SDK coins to Polaris ERC20 tokens
+
+		// mint amount ERC20 tokens to the owner
+		if err = c.callPolarisERC20Mint(sdkCtx, evm, c.RegistryKey(), token, owner, amount); err != nil {
+			return err
+		}
+	}
+
+	// burn amount SDK/Polaris coins from owner
+	if err = cosmlib.BurnCoinsFromAddress(sdkCtx, c.bk, erc20types.ModuleName, owner, denom, amount); err != nil {
+		return err
+	}
+
+	sdkCtx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			erc20types.EventTypeConvertCoinToERC20,
+			sdk.NewAttribute(erc20types.AttributeKeyDenom, denom),
+			sdk.NewAttribute(erc20types.AttributeKeyToken, token.Hex()),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, amount.String()+denom),
+		),
+	)
+	return nil
+}
+
+// convertERC20ToCoin converts ERC20 tokens to SDK/Polaris coins for an owner.
+func (c *Contract) convertERC20ToCoin(
+	ctx context.Context,
+	caller common.Address,
+	evm ethprecompile.EVM,
+	token common.Address,
+	owner common.Address,
+	amount *big.Int,
+) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	// get SDK/Polaris coin denomination pairing with ERC20 token
+	resp, err := c.em.CoinDenomForERC20Address(
+		ctx, &erc20types.CoinDenomForERC20AddressRequest{
+			Token: cosmlib.AddressToAccAddress(token).String(),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	if resp.Denom == "" {
+		// if denomination not found, create new pair with ERC20 token <> Polaris coin denomination
+		resp.Denom = c.em.RegisterERC20CoinPair(sdkCtx, token)
```

Comment:
I think there is a control flow / logic issue here

Reply:
what exactly?